File: src/database/VoucherDatabase.py
import sys
import os
import re
import json
import hashlib
from typing import Any, List, Optional

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..')))
from .SQLiteHelper import SQLiteHelper
from src.database.jsonDb import JsonDB
from ..utils.file import get_relative_path

def get_database_instance() -> SQLiteHelper:
    """Get or create a singleton instance of the SQLiteHelper for the voucher database."""
    if not hasattr(get_database_instance, "_instance") or get_database_instance._db_path != get_relative_path("tmp/voucher_database.sqlite"):
        get_database_instance._instance = SQLiteHelper(get_relative_path("tmp/voucher_database.sqlite"))
        get_database_instance._db_path = get_relative_path("tmp/voucher_database.sqlite")
    return get_database_instance._instance

# Banned voucher codes (normalized, no spaces)
BANNED_VOUCHERS = {
    '1234123412341234'
}

def md5(text: str) -> str:
    """Generate MD5 hash for a given string."""
    return hashlib.md5(text.encode("utf-8")).hexdigest()


def extract_voucher_codes(text: str, output_dir: Optional[str] = None) -> List[str]:
    """
    Extract voucher codes from the given text, optionally outputting debug info to output_dir.
    """
    # Simple regex to find numeric voucher codes
    regex = r"\b\d{4}\s*\d{4}\s*\d{4}\s*\d{4}\b"
    matches = re.findall(regex, text) or []

    # Normalize, filter, and deduplicate
    seen = set()
    result = []
    for code in matches:
        normalized = re.sub(r"\s+", "", code)
        if len(normalized) != 16 or normalized in BANNED_VOUCHERS or normalized in seen:
            continue
        seen.add(normalized)
        result.append(normalized)

    # Debug file writing if requested
    if output_dir:
        try:
            os.makedirs(output_dir, exist_ok=True)
            filename_hash = md5(text)
            with open(os.path.join(output_dir, f"{filename_hash}_text.txt"), "w", encoding="utf-8") as f:
                f.write(text)
            with open(os.path.join(output_dir, f"{filename_hash}_regex.txt"), "w", encoding="utf-8") as f:
                f.write(regex)
            with open(os.path.join(output_dir, f"{filename_hash}_result.json"), "w", encoding="utf-8") as f:
                json.dump(result, f, indent=2, ensure_ascii=False)
        except Exception as e:
            safe_print(f"❌\tError writing debug files: {e}", True)

    return result

def safe_print(message, file=sys.stderr):
    """Safely print messages, handling encoding issues"""
    try:
        print(message, file=file)
        file.flush()
    except UnicodeEncodeError:
        # Fallback to ASCII if Unicode fails
        print(message.encode('ascii', 'ignore').decode('ascii'), file=file)
        file.flush()

def normalize_path(path: str) -> str:
    """Normalize path to Unix-style for consistent database storage"""
    if path:
        return os.path.normpath(path).replace('\\', '/')
    return path

def load_vouchers_from_database(db_helper: SQLiteHelper, image_path: str) -> list:
    """Load vouchers from database

    Args:
        db_helper: SQLiteHelper instance
        image_path: Path to filter by specific image

    Returns:
        List of voucher records with parsed codes
    """
    try:
        # Ensure vouchers table exists
        db_helper.create_table('vouchers', [
            'image_path TEXT PRIMARY KEY',
            'codes TEXT NOT NULL',
            'created_at DATETIME DEFAULT CURRENT_TIMESTAMP'
        ])

        # Normalize the image path for consistent lookup
        normalized_path = normalize_path(image_path)
        records = db_helper.select('vouchers',
                                 where='image_path = ?',
                                 params=(normalized_path,))

        # Parse the comma-separated codes into lists
        vouchers = []
        for record in records:
            # Normalize codes: remove spaces, ensure 16 digits
            codes = [re.sub(r'\s+', '', code) for code in record['codes'].split(',') if len(re.sub(r'\s+', '', code)) == 16]
            voucher_entry = {
                'image_path': record['image_path'],
                'codes': codes,
                'created_at': record['created_at']
            }
            vouchers.append(voucher_entry)

        safe_print(f"📖\tLoaded {len(vouchers)} voucher record(s) from database")
        return vouchers

    except Exception as e:
        safe_print(f"❌\tError loading vouchers from database: {str(e)}")
        return []

def store_voucher_in_database(db_helper: SQLiteHelper, voucher_code: str, image_path: str) -> None:
    """Save found voucher to database"""
    try:
        normalized_path = normalize_path(image_path)
        # Normalize voucher code: remove spaces, ensure 16 digits
        normalized_code = re.sub(r'\s+', '', voucher_code)
        if len(normalized_code) != 16:
            safe_print(f"⚠️\tVoucher code '{voucher_code}' is not 16 digits after normalization, skipping")
            return
        if normalized_code in BANNED_VOUCHERS:
            safe_print(f"⛔\tVoucher code '{normalized_code}' is banned, skipping")
            return

        db_helper.create_table('vouchers', [
            'image_path TEXT PRIMARY KEY',
            'codes TEXT NOT NULL',
            'created_at DATETIME DEFAULT CURRENT_TIMESTAMP'
        ])

        existing = db_helper.select('vouchers',
                                  where='image_path = ?',
                                  params=(normalized_path,))

        if existing:
            current_codes = existing[0]['codes']
            # Normalize all existing codes for comparison
            existing_codes = [re.sub(r'\s+', '', code) for code in current_codes.split(',')]
            if normalized_code not in existing_codes:
                updated_codes = f"{current_codes}, {normalized_code}" if current_codes else normalized_code
                db_helper.update('vouchers',
                               {'codes': updated_codes},
                               'image_path = ?',
                               (normalized_path,))
                safe_print(f"📝\tUpdated voucher record with {normalized_code}")
            else:
                safe_print(f"⚠️\tVoucher {normalized_code} already exists for this image, skipping")
        else:
            voucher_data = {
                'image_path': normalized_path,
                'codes': normalized_code
            }
            db_helper.insert('vouchers', voucher_data)
            safe_print(f"💾\tSaved voucher {normalized_code} to database")

    except Exception as e:
        safe_print(f"❌\tError saving voucher to database: {str(e)}")

def storeVoucherJson(voucherCode: str, imagePath: str) -> None:
    """
    Store a voucher code as JSON for a given image path.
    :param voucherCode: The voucher code to store.
    :param imagePath: The image path associated with the voucher.
    """
    db = JsonDB(os.path.join(os.getcwd(), "tmp", "vouchers"))

    try:
        vouchers = db.load(imagePath) or []
        if not isinstance(vouchers, list):
            vouchers = []
    except FileNotFoundError:
        vouchers = []

    vouchers.append(voucherCode)

    # Ensure no duplicates and normalize
    uniqueVouchers = list({
        re.sub(r"\s+", "", v).strip()
        for v in vouchers
    })

    # Filter out banned vouchers and ensure valid length
    filteredVouchers = [
        v for v in uniqueVouchers
        if len(v) == 16 and v not in BANNED_VOUCHERS
    ]

    if not filteredVouchers:
        safe_print(f"⚠️\tNo valid vouchers to store for {imagePath}")
        return

    db.save(imagePath, filteredVouchers)


def loadVoucherJson(imagePath: str) -> Optional[Any]:
    """
    Load a voucher code JSON for a given image path.
    :param imagePath: The image path to load the voucher for.
    :return: The loaded voucher list or None if not found or error.
    """
    db = JsonDB(os.path.join(os.getcwd(), "tmp", "vouchers"))

    try:
        return db.load(imagePath)
    except Exception as e:
        print(f"❌\tError loading voucher JSON: {str(e)}", flush=True)
        return None

File: src/database/VoucherDatabase.js
import path from 'path';
import fs from 'fs-extra';
import JsonDB from './jsonDb.js';

// Banned voucher codes (normalized, no spaces)
const BANNED_VOUCHERS = new Set(['1234123412341234', '1234123422341234']);

/**
 * Extract voucher codes from the given text, optionally outputting debug info to outputDir.
 * @param {string} text
 * @param {string} [outputDir] - Optional directory to write debug files (text, regex, result)
 * @returns {string[]}
 */
export function extractVoucherCodes(text, outputDir) {
  // Simple regex to find alphanumeric codes, adjust as needed for your voucher format
  const regex = /\b\d{4}\s*\d{4}\s*\d{4}\s*\d{4}\b/g;
  const matches = text.match(regex) || [];
  // Normalize: remove all spaces, ensure 16 digits, skip banned vouchers, remove duplicates
  const seen = new Set();
  const result = matches
    .map((code) => code.replace(/\s+/g, ''))
    .filter((code) => {
      if (code.length !== 16 || BANNED_VOUCHERS.has(code) || seen.has(code)) return false;
      seen.add(code);
      return true;
    });

  if (outputDir) {
    try {
      if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
      }
      fs.writeFileSync(path.join(outputDir, 'debug_text.txt'), text, 'utf8');
      fs.writeFileSync(path.join(outputDir, 'debug_regex.txt'), regex.toString(), 'utf8');
      fs.writeFileSync(path.join(outputDir, 'debug_result.json'), JSON.stringify(result, null, 2), 'utf8');
    } catch (e) {
      // Don't throw, just print error
      safePrint(`❌\tError writing debug files: ${e.message}`, true);
    }
  }
  return result;
}

export function safePrint(message, isError = false) {
  try {
    if (isError) {
      process.stderr.write(message + '\n');
    } else {
      process.stdout.write(message + '\n');
    }
  } catch {
    // Fallback to ASCII if Unicode fails
    // eslint-disable-next-line no-control-regex
    const asciiMsg = message.replace(/[^\u0001-\u007F]/g, '');
    if (isError) {
      process.stderr.write(asciiMsg + '\n');
    } else {
      process.stdout.write(asciiMsg + '\n');
    }
  }
}

export function normalizePath(p) {
  if (p) {
    return path.normalize(p).replace(/\\/g, '/');
  }
  return p;
}

/**
 * Load vouchers from database
 * @param {import('./SQLiteHelper.js').default} dbHelper
 * @param {string} imagePath
 * @returns {Promise<Array>}
 */
export async function loadVouchersFromDatabase(dbHelper, imagePath) {
  try {
    // Ensure vouchers table exists
    dbHelper.createTable('vouchers', [
      'image_path TEXT PRIMARY KEY',
      'codes TEXT NOT NULL',
      'created_at DATETIME DEFAULT CURRENT_TIMESTAMP'
    ]);

    const normalizedPath = normalizePath(imagePath);
    const records = dbHelper.select('vouchers', {
      where: 'image_path = ?',
      params: [normalizedPath]
    });

    const vouchers = [];
    for (const record of records) {
      vouchers.push({
        image_path: record.image_path,
        codes: record.codes
          .split(',')
          .map((code) => code.replace(/\s+/g, '').trim())
          .filter((code) => code.length === 16),
        created_at: record.created_at
      });
    }

    safePrint(`📖\tLoaded ${vouchers.length} voucher record(s) from database`);
    return vouchers;
  } catch (e) {
    safePrint(`❌\tError loading vouchers from database: ${e.message}`, true);
    return [];
  }
}

/**
 * Save found voucher to database
 * @param {import('./SQLiteHelper.js').default} dbHelper
 * @param {string} voucherCode
 * @param {string} imagePath
 * @returns {Promise<void>}
 */
export async function storeVoucherInDatabase(dbHelper, voucherCode, imagePath) {
  try {
    const normalizedPath = normalizePath(imagePath);
    // Normalize voucher code: remove spaces, ensure 16 digits
    const normalizedCode = voucherCode.replace(/\s+/g, '');
    if (normalizedCode.length !== 16) {
      safePrint(`⚠️\tVoucher code '${voucherCode}' is not 16 digits after normalization, skipping`);
      return;
    }
    if (BANNED_VOUCHERS.has(normalizedCode)) {
      safePrint(`⛔\tVoucher code '${normalizedCode}' is banned, skipping`);
      return;
    }

    dbHelper.createTable('vouchers', [
      'image_path TEXT PRIMARY KEY',
      'codes TEXT NOT NULL',
      'created_at DATETIME DEFAULT CURRENT_TIMESTAMP'
    ]);

    const existing = await dbHelper.select('vouchers', {
      where: 'image_path = ?',
      params: [normalizedPath]
    });

    if (existing && existing.length > 0) {
      const currentCodes = existing[0].codes;
      // Normalize all existing codes for comparison
      const existingCodes = currentCodes.split(',').map((c) => c.replace(/\s+/g, '').trim());
      if (!existingCodes.includes(normalizedCode)) {
        const updatedCodes = currentCodes ? `${currentCodes}, ${normalizedCode}` : normalizedCode;
        await dbHelper.update('vouchers', { codes: updatedCodes }, 'image_path = ?', [normalizedPath]);
        safePrint(`📝\tUpdated voucher record with ${normalizedCode}`);
      } else {
        safePrint(`⚠️\tVoucher ${normalizedCode} already exists for this image, skipping`);
      }
    } else {
      const voucherData = {
        image_path: normalizedPath,
        codes: normalizedCode
      };
      await dbHelper.insert('vouchers', voucherData);
      safePrint(`💾\tSaved voucher ${normalizedCode} to database`);
    }
  } catch (e) {
    safePrint(`❌\tError saving voucher to database: ${e.message}`, true);
  }
}

/**
 * Store a voucher code as JSON for a given image path.
 * @param {string} voucherCode - The voucher code to store.
 * @param {string} imagePath - The image path associated with the voucher.
 */
export function storeVoucherJson(voucherCode, imagePath) {
  const db = new JsonDB(path.join(process.cwd(), 'tmp/vouchers'));
  /**
   * @type {string[]}
   */
  const vouchers = db.load(imagePath) || [];
  vouchers.push(voucherCode);
  // Ensure no duplicates and normalize
  const uniqueVouchers = Array.from(new Set(vouchers.map((v) => v.replace(/\s+/g, '').trim())));
  // Ensure banned vouchers are not stored
  const filteredVouchers = uniqueVouchers.filter((v) => v.length === 16 && !BANNED_VOUCHERS.has(v));
  if (filteredVouchers.length === 0) {
    safePrint(`⚠️\tNo valid vouchers to store for ${imagePath}`);
    return;
  }
  db.save(imagePath, filteredVouchers);
}

/**
 * Load a voucher code JSON for a given image path.
 * @template T
 * @param {string} imagePath - The image path to load the voucher for.
 * @returns {T|null} The loaded voucher object of type T, or null if not found or error.
 */
export function loadVoucherJson(imagePath) {
  const db = new JsonDB(path.join(process.cwd(), 'tmp/vouchers'));
  try {
    return db.load(imagePath);
  } catch (e) {
    safePrint(`❌\tError loading voucher JSON: ${e.message}`, true);
    return null;
  }
}
